Mage.Sets/src/mage/cards/c/CallForAid.java

```diff
@@ -0,0 +1,140 @@
+        super(card);
+    }
+
+    @Override
+    public CallForAid copy() {
+        return new CallForAid(this);
+    }
+}
+
+
+class CallForAidEffect extends OneShotEffect {
+
+    public CallForAidEffect() {
+        super(Outcome.Benefit);
+        this.staticText = "Gain control of all creatures target opponent controls until end of turn. Untap those creatures. They gain haste until end of turn. You can't attack that player this turn. You can't sacrifice those creatures this turn.";
+    }
+
+    protected CallForAidEffect(CallForAidEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player opponent = game.getPlayer(getTargetPointer().getFirst(game, source));
+        if (opponent == null) {
+            return false;
+        }
+
+        FilterCreaturePermanent filter = new FilterCreaturePermanent();
+        filter.add(new ControllerIdPredicate(opponent.getId()));
+        List<Permanent> opponentCreatures = game.getBattlefield().getActivePermanents(filter, opponent.getId(), game);
+
+        filter = new FilterCreaturePermanent();
+        PermanentReferenceInCollectionPredicate stolenPermanentsPredicate = new PermanentReferenceInCollectionPredicate(opponentCreatures, game);
+        filter.add(stolenPermanentsPredicate);
+        //... gain control of them till end of turn...
+        GainControlAllEffect gainControl = new GainControlAllEffect(Duration.EndOfTurn, filter);
+        gainControl.apply(game, source);
+        //...untap all stolen creatures...
+        new UntapAllEffect(filter).apply(game, source);
+        //...give them haste...
+        GainAbilityAllEffect gainHaste = new GainAbilityAllEffect(HasteAbility.getInstance(), Duration.EndOfTurn, filter);
+        game.addEffect(gainHaste, source);
+        //...make sure the caster cant attack targeted opponent...
+        FilterCreaturePermanent currentCreaturesFilter = new FilterCreaturePermanent();
+        // At the time this filter it built we do not yet control the creatures of target opponent. So we have to combine the predicates.
+        currentCreaturesFilter.add(Predicates.or(new ControllerIdPredicate(source.getControllerId()), stolenPermanentsPredicate));
+
+        GainAbilityAllEffect cantAttackOwnerEffect = new GainAbilityAllEffect(new SimpleStaticAbility(new CantAttackCertainOpponentEffect(opponent)), Duration.EndOfTurn, currentCreaturesFilter);
+        game.addEffect(cantAttackOwnerEffect, source);
+        //...and finally prevent them from being sacrificed.
+        GainAbilityAllEffect cantBeSacrificed = new GainAbilityAllEffect(new SimpleStaticAbility(new CantBeSacrificedSourceEffect()), Duration.EndOfTurn, filter);
+        game.addEffect(cantBeSacrificed, source);
+
+        return true;
+    }
+
+    @Override
+    public CallForAidEffect copy() {
+        return new CallForAidEffect(this);
+    }
+}
+
+class CantAttackCertainOpponentEffect extends RestrictionEffect {
```

Comment:
This is closer, but still not correct - if a player casts Call for Aid and then plays a creature with haste, that creature won't be affected but still needs to be prevented from attacking that opponent.

So the restriction effect needs to consider the opponent and the controller, but not any of the creatures.

Reply:
I think I figured it out. Now it's a global effect which lasts until end of turn and applies to permanents controlled by the caster of "Call for Aid".

Reply:
Don't forget to check control changing after card cast. If opponent return control of the creature then it must able to do all actions with it like sacrifice. Current code will keep sacrifice restriction instead.